GVFS/GVFS.Common/Git/GitSsl.cs

```diff
@@ -0,0 +1,208 @@
+                        certificates.Select(x => x.Subject));
+        }
+
+        private X509Certificate2 GetCertificateFromFile(ITracer tracer, EventMetadata metadata, GitProcess gitProcess)
+        {
+            string certificatePassword = null;
+            if (this.isCertificatePasswordProtected)
+            {
+                certificatePassword = this.GetCertificatePassword(tracer, gitProcess);
+
+                if (string.IsNullOrEmpty(certificatePassword))
+                {
+                    tracer.RelatedWarning(
+                        new EventMetadata
+                        {
+                                { "SslCertificate", this.certificatePathOrSubjectCommonName }
+                        },
+                        "Git config indicates, that certificate is password protected, but retrieved password was null or empty!");
+                }
+
+                metadata.Add("isPasswordSpecified", string.IsNullOrEmpty(certificatePassword));
+            }
+
+            if (File.Exists(this.certificatePathOrSubjectCommonName))
+            {
+                try
+                {
+                    X509Certificate2 cert = new X509Certificate2(this.certificatePathOrSubjectCommonName, certificatePassword);
+                    if (this.ShouldVerify && cert != null && !cert.Verify())
+                    {
+                        tracer.RelatedWarning(metadata, "Certficate was found, but is invalid.");
+                        return null;
+                    }
+
+                    return cert;
+                }
+                catch (CryptographicException cryptEx)
+                {
+                    metadata.Add("Exception", cryptEx);
+                    tracer.RelatedError(metadata, "Error, while loading certificate from disk");
+                    return null;
+                }
+            }
+
+            return null;
+        }
+
+        private X509Certificate2 GetCertificateFromStore(ITracer tracer, EventMetadata metadata)
+        {
+            try
+            {
+                using (X509Store store = new X509Store())
+                {
+                    store.Open(OpenFlags.OpenExistingOnly | OpenFlags.ReadOnly);
+
+                    X509Certificate2Collection findResults = store.Certificates.Find(X509FindType.FindBySubjectName, this.certificatePathOrSubjectCommonName, this.ShouldVerify);
+                    if (findResults?.Count > 0)
+                    {
+                        LogWithAppropriateLevel(
+                            tracer,
+                            metadata,
+                            findResults.OfType<X509Certificate2>(),
+                            string.Format(
+                                "Found {0} certificates by provided name. Matching DNs: {1}",
+                                findResults.Count,
+                                GetSubjectNameLineForLogging(findResults.OfType<X509Certificate2>())));
```

Comment:
Will you break this single line up and use some temp variables for better readability?  Like
```
string subjects = GetSubjectNameLineForLogging(findResults.OfType<X509Certificate2>());
string logMessage = $"Found {findResults.Count} certificates by provided name. Matching DNs: {subjects}";
LogWithAppropriateLevel(tracer, metadata, findResults.OfType<X509Certificate2>(), logMessage);
```

Reply:
Unless you insist - I'd prefer not to. Adding these temp variables will just pollute the code IMO.

Reply:
@wilbaker What are you thoughts?

Reply:
I'm also not a fan of the multi-line indent, but I would recommend the following:

- Push the formatting work down to the helper method
- Put all of the interesting data into an `EventMetadata`
- Remove `GetSubjectNameLineForLogging `

*Example calling code*

```
LogWithAppropriateLevel(
    tracer, 
    metadata, 
    findResults.OfType<X509Certificate2>(), 
    "Found certificates by provided name");
```

And then move the certificate related formatting into `LogWithAppropriateLevel`

```
private static void LogWithAppropriateLevel(ITracer tracer, EventMetadata metadata, IEnumerable<X509Certificate2> certificates, string logMessage)
{
    int numberOfCertificates = certificates.Count();

    EventMetadata certificateMetadata = new EventMetadata(metadata);
    certificateMetadata.Add("CertificateCount", numberOfCertificates);
    certificateMetadata.Add("CertificateDNs", string.Join(Environment.NewLine, certificates.Select(x => x.Subject)));

    switch (numberOfCertificates)
    {
        case 0:
            tracer.RelatedError(certificateMetadata, logMessage);
            break;
        case 1:
            tracer.RelatedInfo(certificateMetadata, logMessage);
            break;
        default:
            tracer.RelatedWarning(certificateMetadata, logMessage);
            break;
    }
}
```

@kewillford what do you think?